Core/clim-basic/windowing/ports.lisp

```diff
@@ -43,34 +43,34 @@
             (t :clx-ttf))
     :null))
 
-(defun find-default-server-path ()
-  (loop for port in *server-path-search-order*
-        if (get port :port-type)
-           do (return-from find-default-server-path (list port))
-        finally (error "No CLIM backends have been loaded!")))
+(defgeneric find-port-type (port)
+  (:method ((port symbol))
+    (values (get port :port-type)
+            (get port :server-path-parser))))
 
 (defvar *all-ports* nil)
 
 (defun find-port (&key (server-path *default-server-path*))
-  (if (null server-path)
-      (setq server-path (find-default-server-path)))
-  (if (atom server-path)
-      (setq server-path (list server-path)))
-  (setq server-path
-        (funcall (get (first server-path) :server-path-parser) server-path))
-  (loop for port in *all-ports*
-        if (equal server-path (port-server-path port))
-        do (return port)
-        finally (let ((port-type (get (first server-path) :port-type))
-                      port)
-                  (if (null port-type)
-                      (error "Don't know how to make a port of type ~S"
-                             server-path))
-                  (setq port
-                        (funcall 'make-instance port-type
-                                 :server-path server-path))
-                  (push port *all-ports*)
-                  (return port))))
+  (setf server-path (alexandria:ensure-list server-path))
+  (let (port-class)
+    (flet ((try-port (type path)
+             (multiple-value-bind (class parser) (find-port-type type)
+               (when class
+                 (if (null parser)
+                     (setf port-class class
+                           server-path path)
+                     (setf port-class class
+                           server-path (funcall parser path)))))))
```

Comment:
Since the behavior for `class` does not depend on `parser`, maybe like this?
```
(when class
  (setf port-class class
        server-path (if (null parser) 
                        path
                        (funcall parser path))))

Reply:
done.